### PR TITLE
 Add MongoDB and Redis user provider

### PR DESCRIPTION
**Pull request by lpatte**
- MongoDB User Lifecycle
- Redis User Lifecycle
- Import Test for Cloud Databases resources
- Fix Database User Resource Import
- Add ForceNew in Database Resource
- add Retry in IP restriction to work with MongoDB 
- Use ValidateFunc for Database User engine input instead custom validation process

### CODE REVIEW

**Comment by lpatte** [comment]
@ovh/cloud-databases please review

File: ovh/resource_cloud_project_database_mongodb_user.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	params := (&CloudProjectDatabaseMongodbUserUpdateOpts{}).FromResource(d)
+
+	log.Printf("[DEBUG] Will update user: %+v from cluster %s from project %s", params, clusterId, serviceName)
+	err := config.OVHClient.Put(endpoint, params, nil)
+	if err != nil {
+		return fmt.Errorf("calling Put %s with params %s:\n\t %q", endpoint, params, err)
+	}
+
+	log.Printf("[DEBUG] Waiting for database %s to be READY", clusterId)
+	err = waitForCloudProjectDatabaseReady(config.OVHClient, serviceName, "mongodb", clusterId, 60*time.Second, 10*time.Second)
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_mongodb_user.go:186]: Here you should wait for the user to be properly created waiting for the status to be set at "READY" Same goes for other user engines

File: ovh/resource_cloud_project_database_mongodb_user.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	res := &CloudProjectDatabaseMongodbUserResponse{}
+
+	log.Printf("[DEBUG] Will create user: %+v for cluster %s from project %s", params, clusterId, serviceName)
+	err := config.OVHClient.Post(endpoint, params, res)
+	if err != nil {
+		return fmt.Errorf("calling Post %s with params %s:\n\t %q", endpoint, params, err)
+	}
+
+	log.Printf("[DEBUG] Waiting for database %s to be READY", clusterId)
+	err = waitForCloudProjectDatabaseReady(config.OVHClient, serviceName, "mongodb", clusterId, 60*time.Second, 10*time.Second)
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_mongodb_user.go:124]: Here you should wait for the user to be properly created waiting for the status to be set at "READY" Same goes for other user engines

File: ovh/resource_cloud_project_database_mongodb_user.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+				Required:    true,
+			},
+			"roles": {
+				Type:        schema.TypeList,
+				Description: "Roles the user belongs to (without authentication database)",
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+
+			//Optional/Computed
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_mongodb_user.go:51]: User should not be able to set it's password. This endpoint will be deprecated, we don't want user to get use to use this "feature"

File: ovh/resource_cloud_project_database_redis_user.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	res := &CloudProjectDatabaseRedisUserResponse{}
+
+	log.Printf("[DEBUG] Will create user: %+v for cluster %s from project %s", params, clusterId, serviceName)
+	err := config.OVHClient.Post(endpoint, params, res)
+	if err != nil {
+		return fmt.Errorf("calling Post %s with params %+v:\n\t %q", endpoint, params, err)
+	}
+
+	log.Printf("[DEBUG] Waiting for database %s to be READY", clusterId)
+	err = waitForCloudProjectDatabaseReady(config.OVHClient, serviceName, "redis", clusterId, 60*time.Second, 10*time.Second)
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_redis_user.go:128]: Here you should wait for the user to be properly created waiting for the status to be set at "READY"

File: ovh/resource_cloud_project_database_redis_user.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	)
+	params := (&CloudProjectDatabaseRedisUserUpdateOpts{}).FromResource(d)
+
+	log.Printf("[DEBUG] Will update user: %+v from cluster %s from project %s", params, clusterId, serviceName)
+	err := config.OVHClient.Put(endpoint, params, nil)
+	if err != nil {
+		return fmt.Errorf("calling Put %s with params %+v:\n\t %q", endpoint, params, err)
+	}
+
+	log.Printf("[DEBUG] Waiting for database %s to be READY", clusterId)
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_redis_user.go:189]: Here you should wait for the user to be properly created waiting for the status to be set at "READY"

File: ovh/resource_cloud_project_database_mongodb_user.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+		url.PathEscape(id),
+	)
+
+	log.Printf("[DEBUG] Will delete useruser %s from cluster %s from project %s", id, clusterId, serviceName)
+	err := config.OVHClient.Delete(endpoint, nil)
+	if err != nil {
+		return helpers.CheckDeleted(d, err, endpoint)
+	}
+
+	log.Printf("[DEBUG] Waiting for database %s to be DELETED", id)
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_mongodb_user.go:210]: Message shoud be "[DEBUG] Waiting for user %s to be DELETED"

File: ovh/resource_cloud_project_database_postgresql_user.go [hunk 7]
```diff
@@ -195,12 +196,12 @@ func resourceCloudProjectDatabasePostgresqlUserDelete(d *schema.ResourceData, me
 		return helpers.CheckDeleted(d, err, endpoint)
 	}
 
-	log.Printf("[DEBUG] Waiting for database %s to be READY", clusterId)
-	err = waitForCloudProjectDatabaseReady(config.OVHClient, serviceName, "postgresql", clusterId, 20*time.Second, 1*time.Second)
+	log.Printf("[DEBUG] Waiting for database %s to be DELETED", id)
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_postgresql_user.go:199]: message should be" [DEBUG] Waiting for user %s to be DELETED"

File: ovh/resource_cloud_project_database_redis_user.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+		url.PathEscape(id),
+	)
+
+	log.Printf("[DEBUG] Will delete useruser %s from cluster %s from project %s", id, clusterId, serviceName)
+	err := config.OVHClient.Delete(endpoint, nil)
+	if err != nil {
+		return helpers.CheckDeleted(d, err, endpoint)
+	}
+
+	log.Printf("[DEBUG] Waiting for database %s to be DELETED", id)
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_redis_user.go:217]: Message shoud be "[DEBUG] Waiting for user %s to be DELETED"

File: ovh/resource_cloud_project_database_user.go [hunk 9]
```diff
@@ -189,12 +190,12 @@ func resourceCloudProjectDatabaseUserDelete(d *schema.ResourceData, meta interfa
 		return helpers.CheckDeleted(d, err, endpoint)
 	}
 
-	log.Printf("[DEBUG] Waiting for database %s to be READY", clusterId)
-	err = waitForCloudProjectDatabaseReady(config.OVHClient, serviceName, engine, clusterId, 20*time.Second, 1*time.Second)
+	log.Printf("[DEBUG] Waiting for database %s to be DELETED", id)
```
Comment (BellionBastien) [inline ovh/resource_cloud_project_database_user.go:193]: message should be "[DEBUG] Waiting for user %s to be DELETED"

**Review by BellionBastien — APPROVED** [review]

**Comment by scraly** [comment]
review in progress

one modification to do in the doc please:

```
resource "ovh_cloud_project_database" "mongodb" {
  service_name = var.openstack_infos.project_id
```
 
 ```
 resource "ovh_cloud_project_database" "mongodb" {
  service_name = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

:)

File: website/docs/d/cloud_project_database_redis_user.html.markdown [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+* `cluster_id` - Cluster ID
+
+* `name` - Name of the user
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `categories` - Categories of the user.
+* `channels` - Channels of the user.
```
Comment (scraly) [inline website/docs/d/cloud_project_database_redis_user.html.markdown:41]:
```suggestion
* `channels` - Channels of the user.
* `cluster_id` - ID of the cluster.
```

File: ovh/data_cloud_project_database_mongodb_user_test.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+data "ovh_cloud_project_database_mongodb_user" "user" {
+  service_name = ovh_cloud_project_database_mongodb_user.user.service_name
+  cluster_id   = ovh_cloud_project_database_mongodb_user.user.cluster_id
+  name     = ovh_cloud_project_database_mongodb_user.user.name
+}
+`
+
+func TestAccCloudProjectDatabaseMongodbUserDataSource_basic(t *testing.T) {
+	serviceName := os.Getenv("OVH_CLOUD_PROJECT_SERVICE_TEST")
+	version := os.Getenv("OVH_CLOUD_PROJECT_DATABASE_POSTGRESQL_VERSION_TEST")
```
Comment (scraly) [inline ovh/data_cloud_project_database_mongodb_user_test.go:39]:
```suggestion
	version := os.Getenv("OVH_CLOUD_PROJECT_DATABASE_MONGODB_VERSION_TEST")
```

File: ovh/data_cloud_project_database_redis_user_test.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+data "ovh_cloud_project_database_redis_user" "user" {
+  service_name = ovh_cloud_project_database_redis_user.user.service_name
+  cluster_id   = ovh_cloud_project_database_redis_user.user.cluster_id
+  name     = ovh_cloud_project_database_redis_user.user.name
+}
+`
+
+func TestAccCloudProjectDatabaseRedisUserDataSource_basic(t *testing.T) {
+	serviceName := os.Getenv("OVH_CLOUD_PROJECT_SERVICE_TEST")
+	version := os.Getenv("OVH_CLOUD_PROJECT_DATABASE_POSTGRESQL_VERSION_TEST")
```
Comment (scraly) [inline ovh/data_cloud_project_database_redis_user_test.go:42]:
```suggestion
	version := os.Getenv("OVH_CLOUD_PROJECT_DATABASE_REDIS_VERSION_TEST")
```

File: ovh/import_cloud_project_database_redis_user_test.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccCloudProjectDatabaseRedisUser_importBasic(t *testing.T) {
+	serviceName := os.Getenv("OVH_CLOUD_PROJECT_SERVICE_TEST")
+	version := os.Getenv("OVH_CLOUD_PROJECT_DATABASE_POSTGRESQL_VERSION_TEST")
```
Comment (scraly) [inline ovh/import_cloud_project_database_redis_user_test.go:14]:
```suggestion
	version := os.Getenv("OVH_CLOUD_PROJECT_DATABASE_REDIS_VERSION_TEST")
```

**Review by scraly — CHANGES_REQUESTED** [review]

File: website/docs/d/cloud_project_database_mongodb_user.html.markdown [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+* `username` - Name of the user with the authentication database in the format name@authDB
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `created_at` - Date of the creation of the user.
+* `id` - ID of the user.
+* `name` - Name of the user.
```
[the remark this answers is not in the document]
Comment (lpatte) [reply website/docs/d/cloud_project_database_mongodb_user.html.markdown:42]: password is not exported in datasource, only in resource

**Comment by scraly** [comment]
Thanks, it's OK for me :)

**Pull request merged**